File: agent/droplet_agent/mesh/files.py
"""Files sent directly: offers, serving them with Range, and downloading with resume.

The sender offers `{"t":"offer","id","name","size","mime"}` over the link. The
receiver fetches `GET https://<sender>:<port>/mesh/files/<id>` over the same
mutual TLS, streamed, and resumes an interrupted download with
`Range: bytes=<what it has>-`. Only the peer the offer was made to can fetch
it. When the file is complete and its size is right, the receiver answers
`{"t":"ack","id"}`; if it won't take it at all, `{"t":"nack","id","error"}`.

A partial download is kept as a hidden `.part` file beside the downloads,
keyed by the sender's fingerprint and the offer id, so a re-offer of the same
file (after either side restarted) carries on where it stopped.
"""

from __future__ import annotations

import http.client
import json
import logging
import os
import re
import shutil
import threading
import time
from pathlib import Path

from .tlsctx import client_context

log = logging.getLogger("droplet_agent.mesh.files")

OFFER_ID = re.compile(r"^[0-9a-f]{16,64}$")
CHUNK = 256 * 1024
MAX_SIZE = 1 << 40          # 1 TiB: anything bigger is a lie
SPACE_MARGIN = 64 * 1024 * 1024
DOWNLOAD_TRIES = 5
KEEP_COMPLETED = 500


def safe_name(raw) -> str:
    """A file name that can't escape the downloads folder or hide itself."""
    name = str(raw or "").replace("\\", "/").split("/")[-1]
    name = "".join(ch for ch in name if ch.isprintable() and ch not in '<>:"|?*').strip()
    name = name.lstrip(".").strip()
    if len(name.encode()) > 200:
        stem, dot, ext = name.rpartition(".")
        ext = ext if dot and len(ext) <= 16 else ""
        base = (stem if dot else name).encode()[:200 - len(ext) - 1].decode(errors="ignore")
        name = f"{base}.{ext}" if ext else base
    return name or "file"


def unique_path(directory: Path, name: str) -> Path:
    p = directory / name
    stem, suffix = p.stem, p.suffix
    i = 1
    while p.exists():
        p = directory / f"{stem} ({i}){suffix}"
        i += 1
    return p


def parse_range(header: str | None, size: int) -> tuple[int, int] | None | str:
    """A `Range` header → (first, last) inclusive; None for the whole file; "bad" if unsatisfiable.

    Only a single `bytes=` range is supported, which is all resuming needs.
    """
    if not header:
        return None
    m = re.fullmatch(r"\s*bytes\s*=\s*(\d*)\s*-\s*(\d*)\s*", header)
    if not m or (m.group(1) == "" and m.group(2) == ""):
        return None   # a form we don't support: send it all, as HTTP allows
    if m.group(1) == "":
        n = int(m.group(2))            # the last n bytes
        if n == 0 or size == 0:
            return "bad"
        return max(0, size - n), size - 1
    first = int(m.group(1))
    last = int(m.group(2)) if m.group(2) else size - 1
    if first >= size or last < first:
        return "bad"
    return first, min(last, size - 1)


class Offer:
    def __init__(self, oid: str, fp: str, name: str, size: int, mime: str, opener, check=None):
        self.id, self.fp, self.name, self.size, self.mime = oid, fp, name, size, mime
        self.opener = opener          # () -> a binary file object
        self.check = check            # () -> None, or why the file can't be sent any more
        self.sent = 0                 # bytes served, all requests together
        self.last_activity = time.monotonic()
        self.done = threading.Event()
        self.result: tuple[bool, str] | None = None   # (ok, error)

    def message(self) -> dict:
        return {"t": "offer", "id": self.id, "name": self.name, "size": self.size, "mime": self.mime}

    def finish(self, ok: bool, error: str = ""):
        if self.result is None:
            self.result = (ok, error)
        self.done.set()


class Offers:
    """What this device is offering, by id."""

    def __init__(self, max_rate: int = 0):
        self._lock = threading.Lock()
        self._offers: dict[str, Offer] = {}
        self.max_rate = max_rate   # bytes a second when serving, 0 for no limit

    def add(self, offer: Offer):
        with self._lock:
            self._offers[offer.id] = offer

    def get(self, oid: str) -> Offer | None:
        with self._lock:
            return self._offers.get(oid)

    def remove(self, oid: str):
        with self._lock:
            self._offers.pop(oid, None)

    def resolve(self, fp: str, oid: str, ok: bool, error: str = ""):
        """The receiver's ack or nack. Only the peer the offer was made to counts."""
        o = self.get(oid)
        if o is not None and o.fp == fp:
            o.finish(ok, error)

    def serve(self, sock, oid: str, fp: str, headers: dict, method: str, send_head):
        """Answer GET/HEAD /mesh/files/<oid> for the peer `fp`. `send_head(status, headers)`."""
        o = self.get(oid) if OFFER_ID.match(oid or "") else None
        if o is None or o.fp != fp:
            # the same answer whether it doesn't exist or isn't for you
            send_head(404, {"Content-Length": "0"})
            return
        why = o.check() if o.check else None
        if why:
            send_head(410, {"Content-Length": "0"})
            o.finish(False, why)
            return
        rng = parse_range(headers.get("range"), o.size)
        if rng == "bad":
            send_head(416, {"Content-Range": f"bytes */{o.size}", "Content-Length": "0"})
            return
        first, last = (0, o.size - 1) if rng is None else rng
        length = max(0, last - first + 1)
        head = {"Content-Type": o.mime or "application/octet-stream", "Content-Length": str(length),
                "Accept-Ranges": "bytes"}
        if rng is not None:
            head["Content-Range"] = f"bytes {first}-{last}/{o.size}"
        send_head(206 if rng is not None else 200, head)
        if method == "HEAD" or length == 0:
            return
        o.last_activity = time.monotonic()
        started = time.monotonic()
        done = 0
        with o.opener() as f:
            f.seek(first)
            while done < length:
                data = f.read(min(CHUNK, length - done))
                if not data:
                    raise OSError("the file got shorter while it was being sent")
                sock.sendall(data)
                done += len(data)
                o.sent += len(data)
                o.last_activity = time.monotonic()
                if self.max_rate:
                    ahead = done / self.max_rate - (time.monotonic() - started)
                    if ahead > 0:
                        time.sleep(ahead)


class Completed:
    """Offers already received, by (sender fingerprint, id), so a re-offer isn't saved twice."""

    def __init__(self, path: Path):
        self.path = path
        self._lock = threading.Lock()
        try:
            data = json.loads(path.read_text())
            self._items = [x for x in data if isinstance(x, dict)] if isinstance(data, list) else []
        except (OSError, ValueError):
            self._items = []

    def has(self, fp: str, oid: str) -> str | None:
        with self._lock:
            for x in self._items:
                if x.get("fp") == fp and x.get("id") == oid:
                    return x.get("path") or ""
        return None

    def add(self, fp: str, oid: str, path: str):
        with self._lock:
            self._items.append({"fp": fp, "id": oid, "path": path, "ts": int(time.time())})
            self._items = self._items[-KEEP_COMPLETED:]
            self.path.parent.mkdir(parents=True, exist_ok=True)
            tmp = self.path.with_suffix(".tmp")
            tmp.write_text(json.dumps(self._items))
            tmp.replace(self.path)


class DownloadError(Exception):
    def __init__(self, message: str, permanent: bool = False):
        super().__init__(message)
        self.permanent = permanent


def part_paths(directory: Path, fp: str, oid: str) -> tuple[Path, Path]:
    base = directory / f".droplet-{fp[:16]}-{oid}"
    return base.with_name(base.name + ".part"), base.with_name(base.name + ".json")


def check_offer(msg: dict) -> tuple[str, str, int, str]:
    """(id, safe name, size, mime) from an offer. Raises DownloadError(permanent) if it's malformed."""
    oid, size = msg.get("id"), msg.get("size")
    if not isinstance(oid, str) or not OFFER_ID.match(oid):
        raise DownloadError("bad offer id", True)
    if not isinstance(size, int) or isinstance(size, bool) or not 0 <= size <= MAX_SIZE:
        raise DownloadError("bad size", True)
    mime = msg.get("mime") if isinstance(msg.get("mime"), str) else ""
    mime = mime if re.fullmatch(r"[\w.+-]+/[\w.+-]+", mime or "") else "application/octet-stream"
    return oid, safe_name(msg.get("name")), size, mime


def download(identity, fp: str, hosts: list[tuple[str, int]], oid: str, name: str, size: int,
             directory: Path, *, tries: int = DOWNLOAD_TRIES, timeout: float = 20, sleep=time.sleep,
             on_progress=None) -> Path:
    """Fetch an offer into `directory`, resuming a partial one. Returns the saved file."""
    directory.mkdir(parents=True, exist_ok=True)
    part, meta = part_paths(directory, fp, oid)
    try:
        prev = json.loads(meta.read_text())
    except (OSError, ValueError):
        prev = None
    if not isinstance(prev, dict) or prev.get("size") != size or not part.exists() or part.stat().st_size > size:
        part.unlink(missing_ok=True)
        meta.write_text(json.dumps({"size": size, "name": name, "fp": fp}))
    have = part.stat().st_size if part.exists() else 0
    free = shutil.disk_usage(directory).free
    if size - have + SPACE_MARGIN > free:
        raise DownloadError(f"not enough space for {name} ({size} bytes)", True)
    last_error = "no address to fetch it from"
    delay = 1.0
    for attempt in range(tries):
        for host, port in hosts:
            have = part.stat().st_size if part.exists() else 0
            if have == size:
                break
            try:
                _fetch(identity, fp, host, port, oid, part, have, size, timeout, on_progress)
            except DownloadError as e:
                if e.permanent:
                    raise
                last_error = str(e)
                continue
            except (OSError, http.client.HTTPException) as e:
                last_error = f"{host}: {e}"
                continue
            break
        if part.exists() and part.stat().st_size == size:
            break
        if attempt < tries - 1:
            got = part.stat().st_size if part.exists() else 0
            log.info("download of %s stopped at %d of %d bytes (%s); resuming in %.0f s",
                     name, got, size, last_error, delay)
            sleep(delay)
            delay = min(delay * 2, 16)
    else:
        raise DownloadError(last_error)
    if not part.exists():
        part.touch()
    if part.stat().st_size != size:
        raise DownloadError(f"got {part.stat().st_size} bytes, expected {size}")
    final = _claim(directory, name, part)
    meta.unlink(missing_ok=True)
    return final


def _claim(directory: Path, name: str, part: Path) -> Path:
    """Give the finished part its final, unique name, never overwriting anything."""
    while True:
        final = unique_path(directory, name)
        try:
            os.link(part, final)   # fails if the name was taken meanwhile
        except FileExistsError:
            continue
        except OSError:
            if final.exists():
                continue
            os.replace(part, final)   # a filesystem without hard links
            return final
        part.unlink()
        return final


def _fetch(identity, fp, host, port, oid, part: Path, have: int, size: int, timeout, on_progress):
    ctx = client_context(identity, fp)
    conn = http.client.HTTPSConnection(host, port, context=ctx, timeout=timeout)
    try:
        headers = {"User-Agent": "droplet-agent"}
        if have:
            headers["Range"] = f"bytes={have}-"
        conn.request("GET", f"/mesh/files/{oid}", headers=headers)
        resp = conn.getresponse()
        if resp.status in (404, 410):
            raise DownloadError("the sender doesn't offer it any more", True)
        if have and resp.status == 200:
            have = 0          # it ignored the range: start again
        elif resp.status not in (200, 206):
            raise DownloadError(f"the sender answered {resp.status}")
        if resp.status == 206:
            m = re.fullmatch(r"bytes (\d+)-(\d+)/(\d+)", resp.headers.get("Content-Range") or "")
            if not m or int(m.group(1)) != have or int(m.group(3)) != size:
                raise DownloadError("the sender's range doesn't match the partial file")
        if have:
            log.info("resuming %s at byte %d of %d", part.name, have, size)
        with open(part, "r+b" if part.exists() and have else "wb") as f:
            f.seek(have)
            f.truncate()
            got = have
            while True:
                data = resp.read(CHUNK)
                if not data:
                    break
                if got + len(data) > size:
                    raise DownloadError("the sender sent more than it offered", True)
                f.write(data)
                got += len(data)
                if on_progress:
                    on_progress(got, size)
            f.flush()
            os.fsync(f.fileno())
        if got != size:
            raise DownloadError(f"the connection ended at {got} of {size} bytes")
    finally:
        conn.close()

File: windows-net/src/Droplet.Core/Mesh/Files.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using Droplet.Core.Common;

namespace Droplet.Core.Mesh;

/// <summary>
/// File names that can't escape the downloads folder, hide themselves, or be refused
/// by Windows: no path, no leading dots, no characters Windows forbids, no reserved
/// device names, at most 200 bytes (the extension kept).
/// </summary>
public static class SafeName
{
    static readonly HashSet<string> Reserved = new(StringComparer.OrdinalIgnoreCase)
    {
        "CON", "PRN", "AUX", "NUL", "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
        "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9",
    };

    /// <summary>A safe version of <paramref name="raw"/>; "file" when nothing is left.</summary>
    public static string Of(string? raw)
    {
        var name = (raw ?? "").Replace('\\', '/');
        name = name[(name.LastIndexOf('/') + 1)..];
        var sb = new StringBuilder();
        foreach (var r in name.EnumerateRunes())
        {
            if (Clean.IsPrintable(r) && !"<>:\"|?*".Contains(r.ToString(), StringComparison.Ordinal))
            {
                sb.Append(r.ToString());
            }
        }
        name = sb.ToString().Trim().TrimStart('.').Trim();
        // Windows drops trailing dots and spaces itself, which could merge two names
        name = name.TrimEnd('.', ' ');
        if (Encoding.UTF8.GetByteCount(name) > 200)
        {
            var dot = name.LastIndexOf('.');
            var ext = dot > 0 && name.Length - dot - 1 <= 16 ? name[(dot + 1)..] : "";
            var stem = ext.Length > 0 ? name[..dot] : name;
            var budget = 200 - Encoding.UTF8.GetByteCount(ext) - (ext.Length > 0 ? 1 : 0);
            stem = TruncateUtf8(stem, budget);
            name = ext.Length > 0 ? $"{stem}.{ext}" : stem;
        }
        if (name.Length == 0)
        {
            return "file";
        }
        var bare = name.Split('.')[0].TrimEnd(' ');
        return Reserved.Contains(bare) ? "_" + name : name;
    }

    static string TruncateUtf8(string s, int maxBytes)
    {
        var sb = new StringBuilder();
        var n = 0;
        foreach (var r in s.EnumerateRunes())
        {
            n += r.Utf8SequenceLength;
            if (n > maxBytes)
            {
                break;
            }
            sb.Append(r.ToString());
        }
        return sb.ToString();
    }

    /// <summary>
    /// A path in <paramref name="dir"/> for <paramref name="name"/> that doesn't exist yet:
    /// "photo.jpg", then "photo (1).jpg" and so on, as Explorer does.
    /// </summary>
    public static string Unique(string dir, string name)
    {
        var p = Path.Combine(dir, name);
        var ext = Path.GetExtension(name);
        var stem = Path.GetFileNameWithoutExtension(name);
        if (stem.Length == 0)
        {
            (stem, ext) = (name, "");
        }
        for (var i = 1; File.Exists(p) || Directory.Exists(p); i++)
        {
            p = Path.Combine(dir, $"{stem} ({i}){ext}");
        }
        return p;
    }
}

/// <summary>A <c>Range</c> header, as far as resuming needs (a single <c>bytes=</c> range).</summary>
public static partial class RangeHeader
{
    [GeneratedRegex(@"^\s*bytes\s*=\s*(\d*)\s*-\s*(\d*)\s*$")]
    private static partial Regex Pattern();

    /// <summary>What a range means for a file.</summary>
    public enum Kind
    {
        /// <summary>No range (or a form not supported): the whole file.</summary>
        Whole,

        /// <summary>A satisfiable range.</summary>
        Part,

        /// <summary>Unsatisfiable: 416.</summary>
        Bad,
    }

    /// <summary>Parses a header into (kind, first, last) with last inclusive.</summary>
    public static (Kind Kind, long First, long Last) Parse(string? header, long size)
    {
        if (string.IsNullOrEmpty(header))
        {
            return (Kind.Whole, 0, size - 1);
        }
        var m = Pattern().Match(header);
        if (!m.Success || (m.Groups[1].Length == 0 && m.Groups[2].Length == 0))
        {
            return (Kind.Whole, 0, size - 1); // a form we don't support: send it all, as HTTP allows
        }
        if (!TryParse(m.Groups[1].Value, out var a) | !TryParse(m.Groups[2].Value, out var b))
        {
            return (Kind.Bad, 0, 0);
        }
        if (m.Groups[1].Length == 0)
        {
            // the last n bytes
            if (b == 0 || size == 0)
            {
                return (Kind.Bad, 0, 0);
            }
            return (Kind.Part, Math.Max(0, size - b), size - 1);
        }
        var last = m.Groups[2].Length > 0 ? b : size - 1;
        if (a >= size || last < a)
        {
            return (Kind.Bad, 0, 0);
        }
        return (Kind.Part, a, Math.Min(last, size - 1));
    }

    static bool TryParse(string s, out long v)
    {
        v = 0;
        return s.Length == 0 || long.TryParse(s, NumberStyles.None, CultureInfo.InvariantCulture, out v);
    }
}

/// <summary>A file this device offers one peer (docs/mesh.md §9.5).</summary>
public sealed class Offer
{
    /// <summary>Creates an offer.</summary>
    /// <param name="id">16–64 hex.</param>
    /// <param name="fp">The peer it's offered to.</param>
    /// <param name="name">The name it's offered under.</param>
    /// <param name="size">Its size.</param>
    /// <param name="mime">Its type.</param>
    /// <param name="open">Opens it for reading.</param>
    /// <param name="check">Why it can't be sent any more (it changed or went away), or null.</param>
    public Offer(string id, string fp, string name, long size, string mime, Func<Stream> open, Func<string?>? check = null)
    {
        Id = id;
        Fp = fp;
        Name = name;
        Size = size;
        Mime = mime;
        Open = open;
        Check = check;
    }

    /// <summary>The offer id.</summary>
    public string Id { get; }

    /// <summary>The peer it's for.</summary>
    public string Fp { get; }

    /// <summary>The name.</summary>
    public string Name { get; }

    /// <summary>The size.</summary>
    public long Size { get; }

    /// <summary>The type.</summary>
    public string Mime { get; }

    /// <summary>Opens the file.</summary>
    public Func<Stream> Open { get; }

    /// <summary>Why it can't be sent any more, or null.</summary>
    public Func<string?>? Check { get; }

    long sent;
    long lastActivity = Environment.TickCount64;

    /// <summary>Bytes served, all requests together.</summary>
    public long Sent => Interlocked.Read(ref sent);

    /// <summary>How long since a byte was served.</summary>
    public TimeSpan Idle => TimeSpan.FromMilliseconds(Environment.TickCount64 - Interlocked.Read(ref lastActivity));

    internal void Progress(long n)
    {
        Interlocked.Add(ref sent, n);
        Touch();
    }

    internal void Touch() => Interlocked.Exchange(ref lastActivity, Environment.TickCount64);

    readonly TaskCompletionSource<(bool Ok, string Error)> done = new(TaskCreationOptions.RunContinuationsAsynchronously);

    /// <summary>Completes with the receiver's ack (true) or nack (false, why).</summary>
    public Task<(bool Ok, string Error)> Done => done.Task;

    /// <summary>Records the outcome (the first one counts).</summary>
    public void Finish(bool ok, string error = "") => done.TrySetResult((ok, error));

    /// <summary>The offer message.</summary>
    public JsonObject Message() => new() { ["t"] = "offer", ["id"] = Id, ["name"] = Name, ["size"] = Size, ["mime"] = Mime };
}

/// <summary>Offers already received, by (sender, id), so a re-offer isn't saved twice.</summary>
public sealed class CompletedFiles(string path)
{
    const int Keep = 500;

    readonly Lock gate = new();
    readonly List<JsonObject> items = LoadItems(path);

    static List<JsonObject> LoadItems(string path)
    {
        var text = AtomicFile.TryReadText(path);
        if (text is null)
        {
            return [];
        }
        try
        {
            return JsonNode.Parse(text) is JsonArray a ? a.OfType<JsonObject>().Select(o => (JsonObject)o.DeepClone()).ToList() : [];
        }
        catch (JsonException)
        {
            return [];
        }
    }

    /// <summary>Where the offer was saved, "" when unknown, or null when it wasn't received.</summary>
    public string? Has(string fp, string id)
    {
        lock (gate)
        {
            var hit = items.FirstOrDefault(x => x.Str("fp") == fp && x.Str("id") == id);
            return hit is null ? null : hit.Str("path") ?? "";
        }
    }

    /// <summary>Records a received offer.</summary>
    public void Add(string fp, string id, string savedPath)
    {
        lock (gate)
        {
            items.Add(new JsonObject { ["fp"] = fp, ["id"] = id, ["path"] = savedPath, ["ts"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds() });
            if (items.Count > Keep)
            {
                items.RemoveRange(0, items.Count - Keep);
            }
            var a = new JsonArray();
            foreach (var x in items)
            {
                a.Add(x.DeepClone());
            }
            AtomicFile.WriteText(path, Json.ToText(a));
        }
    }
}

/// <summary>A download failed; <see cref="Permanent"/> ones are refused for good (nack).</summary>
public sealed class DownloadException(string message, bool permanent = false, Exception? inner = null) : Exception(message, inner)
{
    /// <summary>Don't retry: tell the sender with a nack.</summary>
    public bool Permanent { get; } = permanent;
}

/// <summary>Checked fields of an offer message.</summary>
/// <param name="Id">The offer id.</param>
/// <param name="Name">A safe name.</param>
/// <param name="Size">The size.</param>
/// <param name="Mime">The type.</param>
public sealed record OfferInfo(string Id, string Name, long Size, string Mime);

/// <summary>
/// Receiving files: checking offers, and downloading with resume. A partial download is
/// kept as a hidden <c>.droplet-&lt;fp prefix&gt;-&lt;id&gt;.part</c> beside the downloads,
/// so a re-offer of the same file (after either side restarted) carries on where it stopped.
/// </summary>
public static partial class FileReceiver
{
    /// <summary>An offer id: 16–64 lowercase hex.</summary>
    [GeneratedRegex("^[0-9a-f]{16,64}$")]
    public static partial Regex OfferIdPattern();

    [GeneratedRegex(@"^[\w.+-]+/[\w.+-]+$")]
    private static partial Regex MimePattern();

    [GeneratedRegex(@"^bytes (\d+)-(\d+)/(\d+)$")]
    private static partial Regex ContentRangePattern();

    /// <summary>Anything bigger is a lie: 1 TiB.</summary>
    public const long MaxSize = 1L << 40;

    const int Chunk = 256 * 1024;
    const long SpaceMargin = 64L * 1024 * 1024;

    /// <summary>Checks an offer message. Throws a permanent <see cref="DownloadException"/> when it's malformed.</summary>
    public static OfferInfo Check(JsonObject msg)
    {
        ArgumentNullException.ThrowIfNull(msg);
        var id = msg.Str("id");
        if (id is null || !OfferIdPattern().IsMatch(id))
        {
            throw new DownloadException("bad offer id", true);
        }
        if (msg.Int("size") is not { } size || size < 0 || size > MaxSize)
        {
            throw new DownloadException("bad size", true);
        }
        var mime = msg.Str("mime") ?? "";
        mime = MimePattern().IsMatch(mime) ? mime : "application/octet-stream";
        return new OfferInfo(id, SafeName.Of(msg.Str("name")), size, mime);
    }

    /// <summary>The partial file and its metadata for an offer.</summary>
    public static (string Part, string Meta) PartPaths(string dir, string fp, string id)
    {
        var baseName = $".droplet-{fp[..16]}-{id}";
        return (Path.Combine(dir, baseName + ".part"), Path.Combine(dir, baseName + ".json"));
    }

    /// <summary>
    /// Fetches an offer into <paramref name="dir"/>, resuming a partial one, trying each
    /// host in turn. Returns the saved file.
    /// </summary>
    public static async Task<string> DownloadAsync(MeshIdentity identity, string fp, IReadOnlyList<(string Host, int Port)> hosts,
        OfferInfo offer, string dir, Action<string>? log = null, Action<long, long>? progress = null, int tries = 5,
        TimeSpan? timeout = null, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(hosts);
        ArgumentNullException.ThrowIfNull(offer);
        Directory.CreateDirectory(dir);
        var (part, meta) = PartPaths(dir, fp, offer.Id);
        var prev = Json.ParseObject(AtomicFile.TryReadText(meta));
        if (prev?.Int("size") != offer.Size || !File.Exists(part) || new FileInfo(part).Length > offer.Size)
        {
            File.Delete(part);
            AtomicFile.WriteText(meta, Json.ToText(new JsonObject { ["size"] = offer.Size, ["name"] = offer.Name, ["fp"] = fp }));
        }
        long Have() => File.Exists(part) ? new FileInfo(part).Length : 0;
        var free = new DriveInfo(Path.GetFullPath(dir)).AvailableFreeSpace;
        if (offer.Size - Have() + SpaceMargin > free)
        {
            throw new DownloadException($"not enough space for {offer.Name} ({offer.Size} bytes)", true);
        }
        var lastError = "no address to fetch it from";
        var delay = TimeSpan.FromSeconds(1);
        for (var attempt = 0; attempt < tries; attempt++)
        {
            foreach (var (host, port) in hosts)
            {
                if (Have() == offer.Size)
                {
                    break;
                }
                try
                {
                    await FetchAsync(identity, fp, host, port, offer, part, Have(), timeout ?? TimeSpan.FromSeconds(20), log, progress, ct)
                        .ConfigureAwait(false);
                    break;
                }
                catch (DownloadException e) when (!e.Permanent)
                {
                    lastError = e.Message;
                }
                catch (Exception e) when (e is IOException or HttpRequestException or TimeoutException ||
                                          (e is TaskCanceledException && !ct.IsCancellationRequested))
                {
                    lastError = $"{host}: {e.Message}";
                }
            }
            if (Have() == offer.Size)
            {
                break;
            }
            if (attempt < tries - 1)
            {
                log?.Invoke($"download of {offer.Name} stopped at {Have()} of {offer.Size} bytes ({lastError}); resuming in {delay.TotalSeconds:0} s");
                await Task.Delay(delay, ct).ConfigureAwait(false);
                delay = TimeSpan.FromSeconds(Math.Min(delay.TotalSeconds * 2, 16));
            }
            else
            {
                throw new DownloadException(lastError);
            }
        }
        if (!File.Exists(part))
        {
            await File.WriteAllBytesAsync(part, [], ct).ConfigureAwait(false);
        }
        if (Have() != offer.Size)
        {
            throw new DownloadException($"got {Have()} bytes, expected {offer.Size}");
        }
        var final = Claim(dir, offer.Name, part);
        File.Delete(meta);
        return final;
    }

    /// <summary>Gives the finished part its final, unique name, never overwriting anything.</summary>
    static string Claim(string dir, string name, string part)
    {
        while (true)
        {
            var final = SafeName.Unique(dir, name);
            try
            {
                File.Move(part, final, overwrite: false);
                return final;
            }
            catch (IOException) when (File.Exists(final))
            {
                // the name was taken meanwhile: pick the next
            }
        }
    }

    static async Task FetchAsync(MeshIdentity identity, string fp, string host, int port, OfferInfo offer, string part, long have,
        TimeSpan timeout, Action<string>? log, Action<long, long>? progress, CancellationToken ct)
    {
        using var handler = MeshTls.Handler(identity, fp, connectTimeout: timeout);
        using var http = new HttpClient(handler) { Timeout = Timeout.InfiniteTimeSpan };
        using var req = new HttpRequestMessage(HttpMethod.Get, new Uri($"https://{Addresses.HostPort(host, port)}/mesh/files/{offer.Id}"))
        {
            Version = System.Net.HttpVersion.Version11,
            VersionPolicy = HttpVersionPolicy.RequestVersionExact,
        };
        req.Headers.TryAddWithoutValidation("User-Agent", MeshProtocol.UserAgent);
        if (have > 0)
        {
            req.Headers.Range = new System.Net.Http.Headers.RangeHeaderValue(have, null);
        }
        using var headersCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        headersCts.CancelAfter(timeout);
        using var resp = await http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, headersCts.Token).ConfigureAwait(false);
        var status = (int)resp.StatusCode;
        if (status is 404 or 410)
        {
            throw new DownloadException("the sender doesn't offer it any more", true);
        }
        if (have > 0 && status == 200)
        {
            have = 0; // it ignored the range: start again
        }
        else if (status is not (200 or 206))
        {
            throw new DownloadException($"the sender answered {status}");
        }
        if (status == 206)
        {
            var cr = resp.Content.Headers.ContentRange?.ToString() ?? "";
            var m = ContentRangePattern().Match(cr);
            if (!m.Success || long.Parse(m.Groups[1].Value, CultureInfo.InvariantCulture) != have ||
                long.Parse(m.Groups[3].Value, CultureInfo.InvariantCulture) != offer.Size)
            {
                throw new DownloadException("the sender's range doesn't match the partial file");
            }
        }
        if (have > 0)
        {
            log?.Invoke($"resuming {Path.GetFileName(part)} at byte {have} of {offer.Size}");
        }
        await using var body = await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        await using var f = new FileStream(part, have > 0 ? FileMode.OpenOrCreate : FileMode.Create, FileAccess.Write, FileShare.None, Chunk, true);
        f.Position = have;
        f.SetLength(have);
        var got = have;
        var buf = new byte[Chunk];
        while (true)
        {
            // no progress for this long: the transfer stopped
            using var readCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            readCts.CancelAfter(timeout);
            int n;
            try
            {
                n = await body.ReadAsync(buf, readCts.Token).ConfigureAwait(false);
            }
            catch (OperationCanceledException) when (!ct.IsCancellationRequested)
            {
                throw new TimeoutException($"no data for {timeout.TotalSeconds:0} s");
            }
            if (n == 0)
            {
                break;
            }
            if (got + n > offer.Size)
            {
                throw new DownloadException("the sender sent more than it offered", true);
            }
            await f.WriteAsync(buf.AsMemory(0, n), ct).ConfigureAwait(false);
            got += n;
            progress?.Invoke(got, offer.Size);
        }
        await f.FlushAsync(ct).ConfigureAwait(false);
        f.Flush(flushToDisk: true);
        if (got != offer.Size)
        {
            throw new DownloadException($"the connection ended at {got} of {offer.Size} bytes");
        }
    }
}
